### PR TITLE
5186 set up JavaScript test infrastructure

### PR DESCRIPTION
**Pull request by TimKam**
Subject: Set up JavaScript test infrastructure

### Feature or Bugfix
- Maintenance

### Purpose
It would be good to have tests for Sphinx's JavaScript code. The goal of this PR is to set up the necessary infrastructure for this. Later, we can gradually increase test coverage.

### Detail
- Refactor JavaScript code to minimize code in template files
- Configure karma.js and write first test
- Configure travis-CI

### Notes
- It looks strange at first glance, but `package-lock.json` should in fact be committed, see: [npm documentation](https://github.com/npm/npm/blob/v5.0.0/doc/files/package-lock.json.md#description)

### Relates
- #5186

### CODE REVIEW

**Comment by TimKam** [comment]
@tk0miya Can you give feedback? If this PR goes into the right direction, I can continue adding tests.

**Comment by tk0miya** [comment]
It seems there are no test runners for `node_js` in `scripts:` section. Would you like to add them in another PR? If intended, I'm okay to this. Please go ahead.

**Comment by TimKam** [comment]
The default Node.js build script is `npm test`, see [here in the Travis-CI docs](https://docs.travis-ci.com/user/languages/javascript-with-nodejs/#Default-Build-Script). That's why I think it's not necessary. The test runs without it. Or am I missing something?

**Comment by tk0miya** [comment]
Please read the log for node_js job.
https://travis-ci.org/sphinx-doc/sphinx/jobs/406888385

We've already overrided it by travis.yml. So default setting is not used. I think you need to call manually like this:
```
  - if [ $RUN_TOX = false ]; then npm test; fi
```
I know `$RUN_TOX` is not good naming for such case.

**Comment by TimKam** [comment]
Sorry for missing this. Now it works. If you think it's okay to merge, we could do so and I'll add new PRs with more tests every now and then.

**Review by tk0miya — APPROVED** [review]
LGTM! Let's merge this :-)

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,4456 @@
...
+      "version": "7.1.2",
+      "resolved": "https://registry.npmjs.org/glob/-/glob-7.1.2.tgz",
+      "integrity": "sha512-MJTUg1kjuLeQCJ+ccE4Vpa6kKVXkPYJ2mOCQyUuKLcLQsdrMCpBPUi8qVE6+YuaJkozeA9NusTAw3hLr8Xe5EQ==",
+      "dev": true,
+      "requires": {
+        "fs.realpath": "1.0.0",
+        "inflight": "1.0.6",
+        "inherits": "2.0.3",
+        "minimatch": "3.0.4",
+
```
Comment (tk0miya) [inline package-lock.json]: I recieved security alert from github.

>We found a potential security vulnerability in a repository for which you have been granted security alert access.
>
>Known moderate severity security vulnerability detected in hoek < 4.2.1defined in package-lock.json.
>package-lock.json update suggested: hoek ~> 4.2.1.

@TimKam Could you check this please? I don't know about this library so much. But it seems too old.

Comment (TimKam) [reply package-lock.json]: There is a problem with dependencies of [karma.js](https://github.com/karma-runner/karma), the test runner, see: https://github.com/karma-runner/karma/issues/2994. It is not possible to update the dependency in the test setup. However, I am confident the vulnerabilities cannot be exploited in our usage scenario, as the libraries are not used in a security-critical context. I suggest it is okay to wait for Karma to update the dependencies and to up our Karma version subsequently.

Comment (tk0miya) [reply package-lock.json]: Thank you for explanation. I understand. And I agree with you.
Okay, let's keep as is :-)

**Pull request merged**